ILM: searchable snapshot executes before migrate in cold/frozen

andreidan opened the pull request:
This moves the execution of the `searchable_snapshot` action before the
`migrate` action in the `cold` and `frozen` phases for a more efficient
data migration (ie. mounting it as a searchable snapshot directly on the
target tier)

Resolves #68635

andreidan committed:
ILM: searchable snapshot executes before migrate in cold/frozen

This moves the execution of the `searchable_snapshot` action before the
`migrate` action in the `cold` and `frozen` phases for a more efficient
data migration (ie. mounting it as a searchable snapshot directly on the
target tier)

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

andreidan commented:
Note: I'll have a look over the needed documentation changes (if any) in a subsequent PR.

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-searchable-snapshot-before-migrate

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -133,6 +133,11 @@ static boolean shouldInjectMigrateStepForPhase(Phase phase) {
             }
         }
 
+        if (phase.getActions().get(SearchableSnapshotAction.NAME) != null) {
+            // the `searchable_snapshot` action defines migration rules itself, so no need to inject a migrate action
+            return false;
+        }
```
I believe we still want to inject a `migrate` action for the `frozen` phase, so that a searchable snapshot created in the `cold` phase transitions to `frozen` nodes when it reaches the frozen phase, even if there is a searchable snapshot action (because the searchable snapshot action *could* end up being a noop if the settings are the same)

andreidan replied:
I think you're right, good shout!

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java:
```diff
@@ -92,22 +96,33 @@ public boolean isSafeAction() {
+            assert DataTier.validTierName(targetTier) : "invalid data tier name:" + targetTier;
+
+            BranchingStep conditionalSkipActionStep = new BranchingStep(preMigrateBranchingKey, migrationKey, nextStepKey,
+                (index, clusterState) -> skipMigrateAction(phase, clusterState.metadata().index(index)));
```
Can you add some `debug` logging about how we're skipping the migration because of XYZ for this? I think it'd be helpful if someone is trying to diagnose why migrate behaves the way it does.

dakrone reviewed (changes requested):
I left a couple of small comments, otherwise this looks good!

andreidan committed:
Inject a migrate action in the frozen phase

andreidan committed:
Log when skipping the migrate action

andreidan commented:
@elasticmahcine update branch

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-searchable-snapshot-before-migrate

andreidan committed:
Docs: migrate auto-injection relation to searchable snapshots

andreidan commented:
Mentioned the relation between the `migrate` and `searchable_snapshot` actions in the docs https://github.com/elastic/elasticsearch/pull/68861/commits/51ba7c7df84f51332c7e1825b827c4ddab719d4e

andreidan committed:
Change migrate skip name to be more meaningful

andreidan committed:
Fix transition for mounted index from searchable_snapshot to another action

Not that searchable_snapshot can precede other actions in the same phase
(eg. in frozen it is followed by `migrate`) we need to allow the mounted
index to resume executing the ILM policy starting with a step that's part
of a new action (ie. migrate).

This adds support to resume the execution of the mounted index from another
action.

With older versions the execution would resume from the PhaseCompleteStep
as it was the last action in a phase, which was handled as a special case
in the `CopyExecutionStateStep`. This commit generalises the `CopyExecutionStateStep`
to be able to resume from any `StepKey`.

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -245,10 +245,8 @@ public String getSnapshotRepository() {
             client, getRestoredIndexPrefix(mountSnapshotKey), getConcreteStorageType(mountSnapshotKey));
         WaitForIndexColorStep waitForGreenIndexHealthStep = new WaitForIndexColorStep(waitForGreenRestoredIndexKey,
             copyMetadataKey, ClusterHealthStatus.GREEN, getRestoredIndexPrefix(waitForGreenRestoredIndexKey));
-        // a policy with only the cold phase will have a null "nextStepKey", hence the "null" nextStepKey passed in below when that's the
-        // case
```
this is not true anymore. the next step will be the (phase) `complete` one.

andreidan commented:
@dakrone having the `searchable_snapshot` before another action in the same phase surfaced another problem with how we used to copy the ILM execution state from the source index to the mounted one. I addressed this in https://github.com/elastic/elasticsearch/pull/68861/commits/5300b2fa7ec21c1336fd65c856dfa52047c966b3

dakrone reviewed (approved):
LGTM, thanks for changing this!
